### PR TITLE
Create new getStateForSvpClient Bridge Method

### PR DESCRIPTION
**Pull request by apancorb**
## Description

The spend transaction should be signed as any other peg-out, but by the proposed powpeg. For this to happen, we need the proposed pegnatories to know when the spend transaction is waiting for signatures.

This also includes a small fix where the `StateForFederators*` where serializing their corresponding RLP raw data twice.

## Motivation and Context
https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP419.md

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java [hunk 1]
```diff
@@ -460,6 +460,18 @@ public enum BridgeMethods {
         fixedPermission(true),
         CallTypeHelper.ALLOW_STATIC_CALL
     ),
+    GET_STATE_FOR_SVP_CLIENT(
+        CallTransaction.Function.fromSignature(
+                "getStateForSvpClient",
+                new String[]{},
+                new String[]{"bytes"}
+        ),
+        fixedCost(4000L),
```
Comment (apancorb) [inline rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java:469]: Maybe we should review this value later on

Comment (jeremy-then) [reply rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java:469]: I don't think it really matters. This function will always be called as `call`. No gas will be spent really. Unless somebody wants to waste gas and uses `send` instead.

Comment (jeremy-then) [reply rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java:469]: Anyways, `getStateForBtcReleaseClient` has 4000L and `getStateForDebugging` has `3_000_000L`. Since this is similar to `getStateForBtcReleaseClient`, defining 4000L would be ok.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 2]
```diff
@@ -1660,6 +1661,23 @@ public byte[] getStateForBtcReleaseClient() throws IOException {
...
+     * Retrieves the current SVP spend transaction state for the SVP client.
+     * <p>
+     * This method checks if there is an SVP spend transaction waiting for signatures, and if so, it serializes 
+     * the state into RLP format. If no transaction is waiting, it returns an encoded empty RLP list.
+     * </p>
+     *
+     * @return A byte array representing the RLP-encoded state of the SVP spend transaction. If no transaction 
+     *         is waiting, returns an RLP-encoded empty list.
+     */
+    public byte[] getStateForSvpClient() {
```
Comment (jeremy-then) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:1674]: 我喜欢

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java [hunk 3]
```diff
@@ -53,15 +52,13 @@ public SortedMap<Keccak256, BtcTransaction> getRskTxsWaitingForSignatures() {
...
+        return BridgeSerializationUtils.serializeRskTxsWaitingForSignatures(
+            rskTxsWaitingForSignatures);
     }
 
     private static byte[] decodeRlpToMap(byte[] rlpData) {
         Objects.requireNonNull(rlpData);
 
-        RLPList rlpList = (RLPList) RLP.decode2(rlpData).get(0);
-        return rlpList.get(0).getRLPData();
+        return RLP.decode2(rlpData).get(0).getRLPData();
```
Comment (julia-zack) [inline rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:62]: why this change?

File: rskj-core/src/test/java/co/rsk/peg/BridgeTest.java [hunk 4]
```diff
@@ -2259,6 +2259,38 @@ void getStateForBtcReleaseClient(MessageCall.MsgType msgType, ActivationConfig a
...
+                !(msgType.equals(MessageCall.MsgType.CALL) || msgType.equals(MessageCall.MsgType.STATICCALL))) {
+                // Post arrowhead should fail for any msg type != CALL or STATIC CALL
+                assertThrows(VMException.class, () -> bridge.execute(data));
+            } else {
+                bridge.execute(data);
+                verify(bridgeSupportMock).getStateForSvpClient();
+            }
+        } else {
+            // Pre RSKIP419 this method is not enabled, should fail for all message types
+            assertThrows(VMException.class, () -> bridge.execute(data));
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/BridgeTest.java:2290]: not a big fan of tests with ifs. I'd better have different tests for different scenarios

File: rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java (hunk 1 above)
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java:469]: Please leave a TODO in the code and a task in the backlog to review this

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 2 above)
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:1674]: Should we return the serialized state here os return the StateForProposedFederator object and let the Bridge serialize it?

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java [hunk 5]
```diff
@@ -53,15 +52,13 @@ public SortedMap<Keccak256, BtcTransaction> getRskTxsWaitingForSignatures() {
      * @return The RLP-encoded byte array representing the current state.
      */
     public byte[] encodeToRlp() {
-        byte[] serializedRskTxsWaitingForSignatures = 
-            BridgeSerializationUtils.serializeRskTxsWaitingForSignatures(rskTxsWaitingForSignatures);
-        return RLP.encodeList(serializedRskTxsWaitingForSignatures);
+        return BridgeSerializationUtils.serializeRskTxsWaitingForSignatures(
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:55]: Did you validate this? Sure it's not breaking consensus?

File: rskj-core/src/test/java/co/rsk/peg/BridgeTest.java (hunk 4 above)
Comment (marcos-iov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeTest.java:2290]: Agree, too many flows here

File: rskj-core/src/main/java/co/rsk/peg/Bridge.java [hunk 6]
```diff
@@ -639,6 +641,17 @@ public byte[] getStateForBtcReleaseClient(Object[] args) throws VMException {
         }
     }
 
+    public byte[] getStateForSvpClient(Object[] args) throws VMException {
+        logger.trace("getStateForSvpClient");
+
+        try {
+            return bridgeSupport.getStateForSvpClient();
```
Comment (marcos-iov) [inline rskj-core/src/main/java/co/rsk/peg/Bridge.java:648]: Does this throw any exception?

File: rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java (hunk 1 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java:469]: Created task to review this.

File: rskj-core/src/main/java/co/rsk/peg/Bridge.java [hunk 7]
```diff
@@ -639,6 +641,17 @@ public byte[] getStateForBtcReleaseClient(Object[] args) throws VMException {
...
+        } catch (Exception e) {
+            logger.warn("Exception in getStateForSvpClient", e);
```
[the remark this answers is not in the document]
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/Bridge.java:650]: Should we create a new task for this refactor?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 2 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:1674]: This is replicating what `getStateForBtcReleaseClient` and `getStateForDebugging` does, so for sake of consistency I would leave it as this.

File: rskj-core/src/test/java/co/rsk/peg/BridgeTest.java (hunk 4 above)
Comment (apancorb) [reply rskj-core/src/test/java/co/rsk/peg/BridgeTest.java:2290]: I agree this is not the prettiest thing out there. But all methods follow this pattern. This test essentially tries all possible permutations, should we maybe create a new task to refactor all of these tests?

File: rskj-core/src/main/java/co/rsk/peg/Bridge.java (hunk 6 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/Bridge.java:648]: This method does not throw any checked exception. But if it throws an unchecked exception, for whatever reason, we should catch it and throw a `VMException`.

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java (hunk 5 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:55]: Up to my knowledge this shouldn't break consensus since this change affects a read-only method of the Bridge smart contract. I guess if during node sync time this forks we would know if this truly breaks consensus. Maybe we can try syncing a node with this change?

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java (hunk 3 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:62]: This only serializes the list twice creating a 2d serialized array. There is no apparent reason for doing this, maybe there is but not to my knowledge.

File: rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java [hunk 8]
```diff
@@ -460,6 +460,18 @@ public enum BridgeMethods {
         fixedPermission(true),
         CallTypeHelper.ALLOW_STATIC_CALL
     ),
+    GET_STATE_FOR_SVP_CLIENT(
+        CallTransaction.Function.fromSignature(
+                "getStateForSvpClient",
+                new String[]{},
+                new String[]{"bytes"}
+        ),
+        fixedCost(4000L), // TODO: check fixed cost value
```
Comment (julia-zack) [inline rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java:469]: i think there's no need to leave a TODO since we already have the task to address this created. Wdyt?

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java (hunk 5 above)
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:55]: maybe a test? :)

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java (hunk 3 above)
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:62]: i won't change anything serialization-related here, we could break consensus

File: rskj-core/src/test/java/co/rsk/peg/BridgeTest.java (hunk 4 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/BridgeTest.java:2290]: we could leave it in the backlog for no-near future jaja.
But at least for now i wouldn't add a test that we know we wanna refactor later

File: rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java (hunk 8 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/BridgeMethods.java:469]: https://github.com/rsksmart/rskj/pull/2746#discussion_r1772257679

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java (hunk 5 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:55]: For serialization/deserialization tests you can find it here: https://github.com/rsksmart/rskj/pull/2738

Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:55]: ah ok!! seems good to me then

File: rskj-core/src/test/java/co/rsk/peg/BridgeTest.java (hunk 4 above)
Comment (apancorb) [reply rskj-core/src/test/java/co/rsk/peg/BridgeTest.java:2290]: It would be simplier to refactor all of them in the same PR and follow the same pattern now. Don't think is a good idea to introduce a major refactor in this PR.

Comment (nathanieliov) [reply rskj-core/src/test/java/co/rsk/peg/BridgeTest.java:2290]: @apancorb, do you think the refactoring can be done during this phase of the svp project? If so, I agree to refactor all the tests at once. Sometimes, it is a bit challenging refactor test structures with legacy structure in the middle.

File: rskj-core/src/main/java/co/rsk/peg/Bridge.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/main/java/co/rsk/peg/Bridge.java:650]: that sounds good yes. Lets do it that way in this new method and address the refactor in another pr, pointing to master maybe. Wdyt?

File: rskj-core/src/test/java/co/rsk/peg/BridgeTest.java (hunk 4 above)
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/BridgeTest.java:2290]: i agree, i won't do the refactor here, but I'd add the new test the way we want it to be.
We can create a new task to address the refactor later, but i wouldn't continue adding things that we know will need a refactor. (specially because we don't actually know if/when we'll do it)

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 2 above)
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:1674]: Ok, we can agree to keep this way here and then refactor all methods in a new PR then :)

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java (hunk 5 above)
Comment (marcos-iov) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java:55]: I think we should rollback this change, until we are 100% certain this doesn't break consensus. That will require some further analysis

File: rskj-core/src/main/java/co/rsk/peg/Bridge.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/Bridge.java:650]: Sounds good

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java (hunk 5 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java]: Sounds good, rollback done

File: rskj-core/src/main/java/co/rsk/peg/StateForFederator.java (hunk 3 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/StateForFederator.java]: This is just that we were serializing the list twice, with the change we serialized once

**Review by marcos-iov — APPROVED** [review]
